crates/data_components/src/s3_vectors/partition.rs

```diff
@@ -0,0 +1,459 @@
+        let column_name = truncate(&sanitize_column(column_name), COLUMN_NAME_MAX_LENGTH);
+        let partition_value_hash = truncate(
+            &hash_to_hex(&partition_value.to_string()),
+            PARTITION_VALUE_MAX_LENGTH,
+        );
+        let partition_by_hash = truncate(
+            &hash_to_hex(&to_stable_string(partition_by).context(UnparsingExpressionSnafu)?),
+            PARTITION_BY_MAX_LENGTH,
+        );
+        Ok(Self {
+            index_name,
+            column_name,
+            partition_value_hash,
+            partition_by_hash,
+        })
+    }
+
+    /// Format an index name suitable for S3 Vectors
+    #[must_use]
+    pub fn to_index_name(&self) -> String {
+        [
+            self.index_name.clone(),
+            self.column_name.clone(),
+            self.partition_value_hash.clone(),
+            self.partition_by_hash.clone(),
+        ]
+        .join(PARTS_SEPARATOR)
+    }
+
+    pub fn from_index_name(index_name: &str) -> Result<Self, Error> {
+        let parts: Vec<&str> = index_name.split(PARTS_SEPARATOR).collect();
+        let num_parts = parts.len();
+        ensure!(num_parts == 4, IncorrectNumPartsInNameSnafu { num_parts });
+        Ok(Self {
+            index_name: parts[0].to_string(),
+            column_name: parts[1].to_string(),
+            partition_value_hash: parts[2].to_string(),
+            partition_by_hash: parts[3].to_string(),
+        })
+    }
+
+    /// Determines if the partitions come from the same dataset
+    #[must_use]
+    pub fn belongs_with(&self, other: &Self) -> BelongsWith {
+        if self.index_name != other.index_name || self.column_name != other.column_name {
+            return BelongsWith::DifferentDataset;
+        }
+        if self.partition_by_hash != other.partition_by_hash {
+            return BelongsWith::DifferentParitionByExpressions;
+        }
+        BelongsWith::SameDataset
+    }
+}
+
+fn sanitize_column(s: &str) -> String {
+    s.replace('_', "-")
+}
+
+fn validate_index(index: &str) -> Result<(), Error> {
+    let len = index.len();
+    ensure!(
+        len < INDEX_NAME_MAX_LENGTH,
+        InvalidIndexNameLengthSnafu {
+            index: index.to_string(),
+            len
+        }
+    );
+    ensure!(!index.contains('-'), InvalidIndexNameHyphenSnafu);
+
+    Ok(())
+}
+
+fn truncate(s: &str, len: usize) -> String {
+    s.chars().take(len).collect()
+}
+
+fn hash_to_hex(input: &str) -> String {
+    let hash = XxHash64::oneshot(HASH_SEED, input.as_bytes());
+    format!("{hash:x}")
+}
+
+// Provide a stable string representation of the expressions
+fn to_stable_string(exprs: &[Expr]) -> Result<String, DataFusionError> {
+    Ok(exprs
+        .iter()
+        .map(stable_expr_string)
+        .collect::<Result<Vec<String>, DataFusionError>>()?
+        .join(PARTS_SEPARATOR))
+}
+
+#[allow(clippy::too_many_lines)]
+fn stable_expr_string(expr: &Expr) -> Result<String, DataFusionError> {
```

Comment:
Should just return our own Error instead of a DataFusion error here, then you get rid of the current awkward error message.